# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoDecoder.cpp
# Commit: cd30b650aeed
# Full Hash: cd30b650aeed44d1fb256a8aed378d27a3bfe39a
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2023-06-29 15:31:45
# Regressor Bug: 1831451
# File Overlap Count: 1
# Description:
#   Bug 1831451 - Generate a unique Id for both ConfigureMessage and DecoderAgent r=alwu
#   
#   To batter tracking the ControlMessage logs, ConfigureMessage should
#   generate a unique Id for itself and its following DecodeMessages and
#   FlushMessages. Besides, that unique id can be assigned to DecoderAgent
# ==============================================================================

diff -r 6892b88cb4ee -r cd30b650aeed dom/media/webcodecs/VideoDecoder.cpp
--- a/dom/media/webcodecs/VideoDecoder.cpp	Thu Jun 29 04:15:04 2023 +0000
+++ b/dom/media/webcodecs/VideoDecoder.cpp	Thu Jun 29 04:15:04 2023 +0000
@@ -621,10 +621,15 @@
     : public ControlMessage,
       public MessageRequestHolder<DecoderAgent::ConfigurePromise> {
  public:
-  explicit ConfigureMessage(UniquePtr<VideoDecoderConfig>&& aConfig)
-      : ControlMessage(nsPrintfCString(
-            "%s-configuration", NS_ConvertUTF16toUTF8(aConfig->mCodec).get())),
-        mConfig(std::move(aConfig)) {}
+  using Id = DecoderAgent::Id;
+  static constexpr Id NoId = 0;
+  static ConfigureMessage* Create(UniquePtr<VideoDecoderConfig>&& aConfig) {
+    // This needs to be atomic since this can run on the main thread or worker
+    // thread.
+    static std::atomic<Id> sNextId = NoId;
+    return new ConfigureMessage(++sNextId, std::move(aConfig));
+  }
+
   ~ConfigureMessage() = default;
   virtual void Cancel() override { Disconnect(); }
   virtual bool IsProcessing() override { return Exists(); };
@@ -632,7 +637,16 @@
   const VideoDecoderConfig& Config() { return *mConfig; }
   UniquePtr<VideoDecoderConfig> TakeConfig() { return std::move(mConfig); }
 
+  const Id mId;  // A unique id shown in log.
+
  private:
+  ConfigureMessage(Id aId, UniquePtr<VideoDecoderConfig>&& aConfig)
+      : ControlMessage(
+            nsPrintfCString("configure #%d (%s)", aId,
+                            NS_ConvertUTF16toUTF8(aConfig->mCodec).get())),
+        mId(aId),
+        mConfig(std::move(aConfig)) {}
+
   UniquePtr<VideoDecoderConfig> mConfig;
 };
 
@@ -663,8 +677,10 @@
     Maybe<uint64_t> mDuration;
   };
 
-  DecodeMessage(Id aId, UniquePtr<ChunkData>&& aData)
-      : ControlMessage(nsPrintfCString("decode #%zu", aId)),
+  DecodeMessage(Id aId, ConfigureMessage::Id aConfigId,
+                UniquePtr<ChunkData>&& aData)
+      : ControlMessage(
+            nsPrintfCString("decode #%zu (config #%d)", aId, aConfigId)),
         mId(aId),
         mData(std::move(aData)) {}
   ~DecodeMessage() = default;
@@ -690,8 +706,9 @@
       public MessageRequestHolder<DecoderAgent::DecodePromise> {
  public:
   using Id = size_t;
-  FlushMessage(Id aId, Promise* aPromise)
-      : ControlMessage(nsPrintfCString("flush #%zu", aId)),
+  FlushMessage(Id aId, ConfigureMessage::Id aConfigId, Promise* aPromise)
+      : ControlMessage(
+            nsPrintfCString("flush #%zu (config #%d)", aId, aConfigId)),
         mId(aId),
         mPromise(aPromise) {}
   ~FlushMessage() = default;
@@ -763,6 +780,7 @@
       mActiveConfig(nullptr),
       mDecodeQueueSize(0),
       mDequeueEventScheduled(false),
+      mLatestConfigureId(ConfigureMessage::NoId),
       mDecodeCounter(0),
       mFlushCounter(0) {
   MOZ_ASSERT(mErrorCallback);
@@ -841,7 +859,8 @@
   mFlushCounter = 0;
 
   mControlMessageQueue.emplace(
-      UniquePtr<ControlMessage>(new ConfigureMessage(std::move(config))));
+      UniquePtr<ControlMessage>(ConfigureMessage::Create(std::move(config))));
+  mLatestConfigureId = mControlMessageQueue.back()->AsConfigureMessage()->mId;
   LOG("VideoDecoder %p enqueues %s", this,
       mControlMessageQueue.back()->ToString().get());
   ProcessControlMessageQueue();
@@ -868,8 +887,9 @@
   }
 
   mDecodeQueueSize += 1;
-  mControlMessageQueue.emplace(UniquePtr<ControlMessage>(new DecodeMessage(
-      ++mDecodeCounter, MakeUnique<DecodeMessage::ChunkData>(aChunk))));
+  mControlMessageQueue.emplace(UniquePtr<ControlMessage>(
+      new DecodeMessage(++mDecodeCounter, mLatestConfigureId,
+                        MakeUnique<DecodeMessage::ChunkData>(aChunk))));
   LOGV("VideoDecoder %p enqueues %s", this,
        mControlMessageQueue.back()->ToString().get());
   ProcessControlMessageQueue();
@@ -893,8 +913,8 @@
 
   mKeyChunkRequired = true;
 
-  mControlMessageQueue.emplace(
-      UniquePtr<ControlMessage>(new FlushMessage(++mFlushCounter, p)));
+  mControlMessageQueue.emplace(UniquePtr<ControlMessage>(
+      new FlushMessage(++mFlushCounter, mLatestConfigureId, p)));
   LOG("VideoDecoder %p enqueues %s", this,
       mControlMessageQueue.back()->ToString().get());
   ProcessControlMessageQueue();
@@ -1267,7 +1287,7 @@
 
   auto i = CreateVideoInfo(msg->Config());
   if (!CanDecode(msg->Config()) || i.isErr() ||
-      !CreateDecoderAgent(msg->TakeConfig(), i.unwrap())) {
+      !CreateDecoderAgent(msg->mId, msg->TakeConfig(), i.unwrap())) {
     mProcessingMessage.reset();
     DebugOnly<Result<Ok, nsresult>> r =
         Close(NS_ERROR_DOM_MEDIA_NOT_SUPPORTED_ERR);
@@ -1535,7 +1555,8 @@
 // ShutdownpPomise-resolver. In case 2, the entry point is in mWorkerRef's
 // shutting down callback. In case 3, the entry point is in mWorkerRef's
 // shutting down callback.
-bool VideoDecoder::CreateDecoderAgent(UniquePtr<VideoDecoderConfig>&& aConfig,
+bool VideoDecoder::CreateDecoderAgent(DecoderAgent::Id aId,
+                                      UniquePtr<VideoDecoderConfig>&& aConfig,
                                       UniquePtr<TrackInfo>&& aInfo) {
   AssertIsOnOwningThread();
   MOZ_ASSERT(mState == CodecState::Configured);
@@ -1571,7 +1592,7 @@
     mWorkerRef = new ThreadSafeWorkerRef(workerRef);
   }
 
-  mAgent = DecoderAgent::Create(std::move(aInfo));
+  mAgent = MakeRefPtr<DecoderAgent>(aId, std::move(aInfo));
   mActiveConfig = std::move(aConfig);
 
   // ShutdownBlockingTicket requires an unique name to register its own